tests/testthat/test-utility-extra.R

```diff
@@ -0,0 +1,400 @@
+    ref = m4ma::iCones_r(p1, a_front, matrix(, 0, 2), r, objects)
+    res = m4ma::iCones_rcpp(p1, a_front, matrix(, 0, 2), r, objects)
+    testthat::expect_equal(res, ref)
+    testthat::expect_equal(res, NULL)
+  }
+)
+
+testthat::test_that(
+  "Intersecting cones works (when one other pedestrian)", {
+    ref = m4ma::iCones_r(p1, a_front, p2[2, , drop = FALSE], r, objects)
+    res = m4ma::iCones_rcpp(p1, a_front, p2[2, , drop = FALSE], r, objects)
+    testthat::expect_equal(res, ref)
+  }
+)
+
+testthat::test_that(
+  "Intersecting cones works (when no intersection)", {
+    ref = m4ma::iCones_r(p1, 175, p2[-n, , drop = FALSE], r, objects)
+    res = m4ma::iCones_rcpp(p1, 175, p2[-n, , drop = FALSE], r, objects)
+    testthat::expect_equal(res, ref)
+    testthat::expect_equal(res, NULL)
+  }
+)
+
+testthat::test_that(
+  "Intersecting cones works (when one cone intersecting and seen)", {
+    ref = m4ma::iCones_r(p1, a_not_front, p2[3, , drop = FALSE], r, objects)
+    res = m4ma::iCones_rcpp(p1, a_not_front, p2[3, , drop = FALSE], r, objects)
+    testthat::expect_equal(res, ref)
+  }
+)
+
+testthat::test_that(
+  "Intersecting cones works (when one cone intersecting but not seen)", {
+    ref = m4ma::iCones_r(p1, a_not_front,
+                         p2[3, , drop = FALSE], r, objects_occlude)
+    res = m4ma::iCones_rcpp(p1, a_not_front,
+                            p2[3, , drop = FALSE], r, objects_occlude)
+    testthat::expect_equal(res, ref)
+    testthat::expect_equal(res, NULL)
+  }
+)
+
+testthat::test_that(
+  "Intersecting cones works (when all pedestrians are occluded in front)", {
+    ref = m4ma::iCones_r(p1, a_front,
+                         p2[-n, , drop = FALSE], r, objects_occlude)
+    res = m4ma::iCones_rcpp(p1, a_front,
+                            p2[-n, , drop = FALSE], r, objects_occlude)
+    testthat::expect_equal(res, ref)
+  })
+
+testthat::test_that("Intersecting cones to cells transformation works", {
+  # Make test independent of iCones
+  cones = c(rnorm(6))
+  names(cones) = as.character(1:6)
+  v = 0.5
+  ref = m4ma::iCones2Cells_r(cones, v)
+  res = m4ma::iCones2Cells_rcpp(cones, v)
+  testthat::expect_equal(res, ref)
+})
+
+testthat::test_that(
+  "Blocked angle works (when intersecting cones and seen)", {
+    ref = m4ma::blockedAngle_r(n, state, p_pred, objects)
+    res = m4ma::blockedAngle_rcpp(n-1, state, p_pred, objects)
+    testthat::expect_equal(res, if (length(ref) == 0) unname(ref) else ref)
+  }
+)
+
+testthat::test_that(
+  "Blocked angle works (when no intersecting cones)", {
+    a_back = c(175, 175, 175)
+    names(a_back) = nms
+    state_occluded = state
+    state_occluded$a = a_back
+    ref = m4ma::blockedAngle_r(n, state_occluded, p2, objects_occlude)
+    res = m4ma::blockedAngle_rcpp(n-1, state_occluded, p_pred, objects_occlude)
+    testthat::expect_equal(res, if (length(ref) == 0) unname(ref) else ref)
+    testthat::expect_equal(res, numeric(0))

```

Comment:
line 213: again here if the functions in r and c++ are doing the same thing, why removing the comparison with the ref, and add 0? and why 'numeric(0)'? isn't 0 already numeric? See the proof of concept below


>str(0)
 num 0
> str(numeric(0))
 num(0)

Reply:
As above, this test always requires the R and Rcpp implementations to output a numeric vector of length zero. `numeric(0)` does not return `0` but the required numeric vector of length zero, i.e., `length(numeric(0)) == 0`. 

For a reason I don't understand yet, Rcpp always returns a named numeric vector of length zero, which is why I have to unname it before testing it against R.

Reply:
ah ok, it makes sense